packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts

```diff
@@ -1,226 +1,178 @@
+      name: DefaultKeyName.PHONE_NUMBER_PRIVACY,
+    }
+
+    let signature: string
+    if (duplicateRequest && duplicateRequest.signature?.length) {
+      signature = duplicateRequest.signature
+    } else {
+      try {
+        signature = await sign(
+          request.body.blindedQueryPhoneNumber,
+          key,
+          keyProvider,
+          response.locals.logger
+        )
+      } catch (err) {
+        response.locals.logger.error({ err }, 'catch error on signing')
+
+        return errorResult(500, ErrorMessage.SIGNATURE_COMPUTATION_FAILURE, {
+          performedQueryCount: usedQuota,
+          totalQuota: account.pnpTotalQuota,
+        })
+      }
+    }
+
+    if (!duplicateRequest) {
+      await requestService.recordRequest(
+        account.address,
+        request.body.blindedQueryPhoneNumber,
+        signature,
+        ctx
+      )
+      usedQuota++
+    } else {
+      Counters.duplicateRequests.inc()
+      logger.info('Request already exists in db. Will service request without charging quota.')
+      warnings.push(WarningMessage.DUPLICATE_REQUEST_TO_GET_PARTIAL_SIG)
+    }
+
+    // Send Success response
+    response.set(KEY_VERSION_HEADER, key.version.toString())
+    return {
+      status: 200,
+      body: {
+        success: true as true,
+        version: getSignerVersion(),
+        signature,
+        performedQueryCount: usedQuota,
+        totalQuota: account.pnpTotalQuota,
+        warnings,
+      },
+    }
   }
+}
 
-  private async sign(
-    blindedMessage: string,
-    key: Key,
-    session: Session<SignMessageRequest>
-  ): Promise<string> {
-    let privateKey: string
+async function sign(
+  blindedMessage: string,
+  key: Key,
+  keyProvider: KeyProvider,
+  logger: Logger
+): Promise<string> {
+  let privateKey: string
+  return traceAsyncFunction('pnpSign', async () => {
     try {
-      privateKey = await this.keyProvider.getPrivateKeyOrFetchFromStore(key)
+      privateKey = await keyProvider.getPrivateKeyOrFetchFromStore(key)
     } catch (err) {
-      session.logger.info({ key }, 'Requested key version not supported')
-      session.logger.error(err)
+      logger.info({ key }, 'Requested key version not supported')
+      logger.error(err)
       throw new Error(WarningMessage.INVALID_KEY_VERSION_REQUEST)
     }
-    return computeBlindedSignature(blindedMessage, privateKey, session.logger)
-  }
+    return computeBlindedSignature(blindedMessage, privateKey, logger)
+  })
+}
+
+function isValidRequest(
+  request: Request<{}, {}, unknown>
+): request is Request<{}, {}, SignMessageRequest> {
+  return (
+    SignMessageRequestSchema.is(request.body) &&
+    hasValidAccountParam(request.body) &&
+    hasValidBlindedPhoneNumberParam(request.body) &&
+    isBodyReasonablySized(request.body)
+  )
+}
+
+function bypassQuotaForE2ETesting(
+  bypassQuotaPercentage: number,
+  requestBody: SignMessageRequest
+): boolean {
+  const sessionID = Number(requestBody.sessionID)
```

Comment:
The sessionID is intended to enable tracing of a request across services (including the identity sdk / client). Though we're not using this in our monitoring, it could come in handy later and I don't see it doing harm as an optional field. Of course, the short-term reason why it would make sense to keep it is that our mechanism for bypassingQuota during load tests depends on it

Reply:
I'll add a TODO to revisit this